### PR TITLE
feat: fetch file's details by ids

### PR DESCRIPTION
**Pull request by alejandro0619**
File's details  can be now fetched by its id.

### CODE REVIEW

File: src/management/file_details.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+#[async_trait]
+pub trait Details {
+    /// Given a file id retrieves the information within
+    async fn get_file_details<T: Display + Send>(&self, id: T) -> Result<Response>;
+}
+
+#[async_trait]
+impl Details for ImageKit {
+    async fn get_file_details<T: Display + Send>(&self, id: T) -> Result<Response> {
```
Comment (EstebanBorai) [inline src/management/file_details.rs:17]: Checking on https://docs.imagekit.io/api-reference/media-api/get-file-details I can see `fileId` is of type string, why not using `<S: AsRef<str>>`?. `Display` and `Send` are too generic for this purpose.

```suggestion
    async fn get_file_details<S: AsRef<str>>(&self, id: S) -> Result<Response> {
```

By using `<S: AsRef<str>>` we allow users to provide:

- `&str`
- `String`
- `&String`

Which is flexible yet safe enough!

File: src/lib.rs [hunk 2]
```diff
@@ -29,8 +33,22 @@ mod tests {
...
 
+        let detail_result = imagekit.get_file_details(&upload_result.file_id).await;
+        assert!(detail_result.is_ok());
+
         let delete_result = imagekit.delete(upload_result.file_id).await;
 
         assert!(delete_result.is_ok());
     }
+    #[tokio::test]
+    async fn uploads_and_retrieve_information() {
```
Comment (EstebanBorai) [inline src/lib.rs:44]: Nitpick!

```suggestion

    #[tokio::test]
    async fn uploads_and_retrieve_information() {
```

File: src/lib.rs [hunk 3]
```diff
@@ -29,8 +33,22 @@ mod tests {
...
+        let imagekit = ImageKit::from_env().unwrap();
+        let file = File::open("assets/ferris.jpeg").await.unwrap();
+        let upload_file = UploadFile::from(file);
+        let opts = Options::new(upload_file, "ferris");
+        let upload_result = imagekit.upload(opts).await.unwrap();
+
+        let detail_result = imagekit.get_file_details(&upload_result.file_id).await;
```
Comment (EstebanBorai) [inline src/lib.rs:51]:
```suggestion
        let upload_result = imagekit.upload(opts).await.unwrap();
        let detail_result = imagekit.get_file_details(&upload_result.file_id).await;
```

File: src/lib.rs [hunk 4]
```diff
@@ -29,8 +33,22 @@ mod tests {
...
+        assert!(detail_result.is_ok());
```
Comment (EstebanBorai) [inline src/lib.rs:52]: I think we should check certain predictable image details to make sure data comes as expected.

File: src/upload/types.rs [hunk 5]
```diff
@@ -56,7 +56,11 @@ pub struct Response {
     /// Width of the image in pixels (Only for Images)
     pub width: Option<u64>,
     /// In the case of an image, a small thumbnail URL.
-    pub thumbnail_url: String,
+    /// Since the API has not normalized some fields such as thumbnail
+    /// A provided workaround is to attempt the parsing from "thumbnail_url" or "thumbnail"
+    /// CHECK: Could this incur in repeating the same proccess twice so it may have a performance impact?
+    #[serde(alias = "thumbnail")]
+    pub thumbnail_url: Option<String>,
```
Comment (EstebanBorai) [inline src/upload/types.rs:63]: This would introduce breaking changes to current version, is this actually required?
Perhaps we could discuss it in a separate PR? Given that this is not related to fetching file details?

File: src/management/file_details.rs (hunk 1 above)
Comment (alejandro0619) [reply src/management/file_details.rs:17]: Seems to be right, thanks I'll get it fixed

File: src/lib.rs (hunk 4 above)
Comment (alejandro0619) [reply src/lib.rs:52]: Perfect, I'll get a template of a response (image details) and try to test against it.

File: src/upload/types.rs (hunk 5 above)
Comment (alejandro0619) [reply src/upload/types.rs:63]: > This would introduce breaking changes to current version, is this actually required?

As long as I've tested the two endpoints which we provided functionality it happened to not use the exactly the same fields. Meanwhile the upload endpoint uses _thumbnail_url_, the details endpoint uses _thumbnail_ hence I tried to make the response structure more general avoiding problems when deserializing the JSON. However, as you said, it'd better to discuss deeper this topic in a separate PR

File: src/management/file_details.rs (hunk 1 above)
Comment (EstebanBorai) [reply src/management/file_details.rs:17]: Would be cool! But please create a separate PR!

File: src/lib.rs [hunk 6]
```diff
@@ -29,8 +33,22 @@ mod tests {
...
     }
+    #[tokio::test]
+    async fn uploads_and_retrieve_information() {
+        let imagekit = ImageKit::from_env().unwrap();
+        let file = File::open("assets/ferris.jpeg").await.unwrap();
+        let upload_file = UploadFile::from(file);
+        let opts = Options::new(upload_file, "ferris");
+        let upload_result = imagekit.upload(opts).await.unwrap();
+
+        let detail_result = imagekit.get_file_details(upload_result.file_id).await;
```
Comment (EstebanBorai) [inline src/lib.rs:51]:
```suggestion
        let upload_result = imagekit.upload(opts).await.unwrap();
        let detail_result = imagekit.get_file_details(upload_result.file_id).await;
```

**Review by EstebanBorai — APPROVED** [review]

**Comment by EstebanBorai** [comment]
Lets merge it! I think this is related to permissions will investigate

**Comment by EstebanBorai** [comment]
@dararod can you fmt this pls?
